src/client/gui/lib/vm_table/bulk_actions.dart

```diff
@@ -9,6 +9,126 @@ import '../providers.dart';
 import '../vm_action.dart';
 import 'vms.dart';
 
+class ZonesDropdown extends ConsumerWidget {
+  const ZonesDropdown({super.key});
+
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
+    // Pre-fetch zones data
+    final zonesAsync = ref.watch(zonesProvider);
+    final vmInfos = ref.watch(vmInfosProvider);
+    return PopupMenuButton(
+      tooltip: 'Change zone availability',
+      position: PopupMenuPosition.under,
+      child: OutlinedButton(
+        onPressed: null,
+        child: const Text('Zones', style: TextStyle(color: Colors.black)),
+      ),
+      itemBuilder: (context) => zonesAsync.when(
+          loading: () => [
+                const PopupMenuItem(
+                  enabled: false,
+                  child: SizedBox(
+                    width: 20,
+                    height: 20,
+                    child: CircularProgressIndicator(),
+                  ),
+                ),
+              ],
+          error: (_, __) => [
+                const PopupMenuItem(
+                  enabled: false,
+                  child: Text('Error loading zones'),
+                ),
+              ],
+          data: (zones) => [
+                const PopupMenuItem(
+                  enabled: false,
+                  child: Padding(
+                    padding: EdgeInsets.symmetric(horizontal: 4, vertical: 8),
+                    child: Text(
+                      'Availability zones',
+                      style: TextStyle(
+                        fontSize: 24,
+                        fontWeight: FontWeight.w600,
+                        color: Colors.black,
+                      ),
+                    ),
+                  ),
+                ),
+                ...zones.map((zone) {
```

Comment:
After you move the retrieval of `runningCount` to the `_ZoenToggleRow` widget, you could go to a `for (final zone in zones)` instead of `...zones.map`, which will look nicer.